hydra-node/src/Hydra/Chain/Offline.hs

```diff
@@ -0,0 +1,138 @@
+
+  tickForeverAction <- case ledgerGenesisFile of
+    Just filePath -> do
+      Ledger.ShelleyGenesis{sgSystemStart, sgSlotLength, sgEpochLength} <-
+        readJsonFileThrow (parseJSON @(Ledger.ShelleyGenesis StandardCrypto)) filePath
+      let slotLengthNominalDiffTime = Ledger.fromNominalDiffTimeMicro sgSlotLength
+          slotLength = mkSlotLength slotLengthNominalDiffTime
+
+      let interpreter = mkInterpreter $ neverForksSummary sgEpochLength slotLength
+
+      let slotFromUTCTime :: HasCallStack => UTCTime -> Either Consensus.PastHorizonException ChainSlot
+          slotFromUTCTime utcTime = do
+            let relativeTime = toRelativeTime (SystemStart sgSystemStart) utcTime
+            case interpretQuery interpreter (wallclockToSlot relativeTime) of
+              Left pastHorizonEx ->
+                Left pastHorizonEx
+              Right (SlotNo slotNoWord64, _timeSpentInSlot, _timeLeftInSlot) ->
+                Right . ChainSlot . fromIntegral @Word64 @Natural $ slotNoWord64
+
+          slotToUTCTime :: HasCallStack => ChainSlot -> Either Consensus.PastHorizonException UTCTime
+          slotToUTCTime (ChainSlot slotNat) =
+            case interpretQuery interpreter (slotToWallclock . SlotNo . fromIntegral @Natural @Word64 $ slotNat) of
+              Left pastHorizonEx -> Left pastHorizonEx
+              Right (relativeTime, _slotLength) -> pure $ Slotting.fromRelativeTime (SystemStart sgSystemStart) relativeTime
+
+      let nextTick (SlotNo upcomingSlotWord64) = do
+            let upcomingSlotChainSlot = ChainSlot . fromIntegral @Word64 @Natural $ upcomingSlotWord64
+            timeToSleepUntil <- either throwIO pure . slotToUTCTime $ upcomingSlotChainSlot
+            sleepDelay <- diffUTCTime timeToSleepUntil <$> getCurrentTime
+            threadDelay $ nominalDelay sleepDelay
+            callback $
+              Tick
+                { chainTime = timeToSleepUntil
+                , chainSlot = ChainSlot . fromIntegral @Word64 @Natural $ upcomingSlotWord64
+                }
+
+      ChainSlot initialSlotNat <- either throwIO pure =<< fmap slotFromUTCTime getCurrentTime
+      let initialSlot = SlotNo . fromIntegral @Natural @Word64 $ initialSlotNat
+      let tickForever = traverse_ nextTick [initialSlot ..]
+      pure tickForever
+    Nothing -> do
+      let epochInfo@EpochInfo{} = epochInfoPure globals
+          initialSlot = runIdentity $ epochInfoFirst epochInfo 0
+
+          nextTick upcomingSlot = do
+            let timeToSleepUntil = runIdentity $ epochInfoSlotToUTCTime epochInfo systemStart upcomingSlot
+            sleepDelay <- diffUTCTime timeToSleepUntil <$> getCurrentTime
+            threadDelay $ nominalDelay sleepDelay
+            callback $
+              Tick
+                { chainTime = timeToSleepUntil
+                , chainSlot = ChainSlot . fromIntegral @Word64 @Natural $ unSlotNo upcomingSlot
+                }
```

Comment:
I think this can be greatly simplified as there will be no epochs and no differring slot lengths (you use `neverForksSummary`). That is, the current slot can be just computed by lookging at `currentTime` and `startTime` from the genesis shelley and devide that by the `slotLength`. No need for the epoch summary interpreters. For the ticking itself you could do something like

```hs
let tickForeverAction = tick slotLengthNominalDiffTime callback startSlot
```
with

```hs
tick :: (MonadDelay m, MonadTime m) => NominalDiffTime -> (ChainEvent tx -> m ()) -> Natural -> m ()
tick slotLengthNominalDiffTime callback start = do
  go start
 where
  go !n = do
    threadDelay $ realToFrac slotLengthNominalDiffTime
    chainTime <- getCurrentTime
    callback $
      Tick
        { chainTime
        , chainSlot = ChainSlot n
        }
    go $ n + 1
```